datafusion/core/src/execution/session_state.rs

```diff
@@ -976,6 +837,482 @@ impl SessionState {
+    ) -> Self {
+        self.state.aggregate_functions = aggregate_functions;
+        self
+    }
+
+    /// override default window functions with `window_functions`
+    pub fn with_window_functions(
+        mut self,
+        window_functions: HashMap<String, Arc<WindowUDF>>,
+    ) -> Self {
+        self.state.window_functions = window_functions;
+        self
+    }
+
+    /// Registers a [`SerializerRegistry`]
+    pub fn with_serializer_registry(
+        mut self,
+        serializer_registry: Arc<dyn SerializerRegistry>,
+    ) -> Self {
+        self.state.serializer_registry = serializer_registry;
+        self
+    }
+
+    /// override default list of file formats with `file_formats`
+    pub fn with_file_formats(
+        mut self,
+        file_formats: HashMap<String, Arc<dyn FileFormatFactory>>,
+    ) -> Self {
+        self.state.file_formats = file_formats;
+        self
+    }
+
+    /// override the session config with `config`
+    pub fn with_config(mut self, config: SessionConfig) -> Self {
+        self.state.config = config;
+        self
+    }
+
+    /// override default table options with `table_options`
+    pub fn with_table_options(mut self, table_options: TableOptions) -> Self {
+        self.state.table_options = table_options;
+        self
+    }
+
+    /// Adds a new [`ConfigExtension`] to TableOptions
+    pub fn with_table_options_extension<T: ConfigExtension>(
+        mut self,
+        extension: T,
+    ) -> Self {
+        self.state.table_options.extensions.insert(extension);
+        self
+    }
+
+    /// override default execution props with `execution_props`
+    pub fn with_execution_props(mut self, execution_props: ExecutionProps) -> Self {
+        self.state.execution_props = execution_props;
+        self
+    }
+
+    /// override default table factories with `table_factories`
+    pub fn with_table_factories(
+        mut self,
+        table_factories: HashMap<String, Arc<dyn TableProviderFactory>>,
+    ) -> Self {
+        self.state.table_factories = table_factories;
+        self
+    }
+
+    /// override the runtime env with `runtime_env`
+    pub fn with_runtime_env(mut self, runtime_env: Arc<RuntimeEnv>) -> Self {
+        self.state.runtime_env = runtime_env;
+        self
+    }
+
+    /// Registers a [`FunctionFactory`] to handle `CREATE FUNCTION` statements
+    pub fn with_function_factory(
+        mut self,
+        function_factory: Option<Arc<dyn FunctionFactory>>,
+    ) -> Self {
+        self.state.function_factory = function_factory;
+        self
+    }
+
+    /// build a [`SessionState`] with the current configuration
+    pub fn build(mut self) -> SessionState {
+        if self.use_defaults {
+            if self.state.table_factories.is_empty() {
+                self.state.table_factories =
+                    SessionStateDefaults::default_table_factories();
+            }
+            if self.state.expr_planners.is_empty() {
```

Comment:
the semantics of "use_defaults" is effectively ignored if the user has provided their own pieces might be confusing

I think it would be easier to reason about if the defaults got installed as part of `SessionBuilder::with_use_defaults()` (rather than a flag that deferred the installation to the end) -- that way the user could decide if they wanted to modify the default lists or start from an empty list

What I am hoping long term is that we could pass in the defaults as an argument (and implement it as a trait) thus breaking the dependency between SessionState and core

```rust
let state = SessionStateBuilder::new()
  .with_defaults(SessionStateDefaults::new())?
  .build()
```